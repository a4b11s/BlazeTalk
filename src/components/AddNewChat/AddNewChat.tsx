import React, { useState } from 'react';

import {
	Avatar,
	Box,
	Button,
	Dialog,
	DialogActions,
	DialogContent,
	DialogTitle,
	Slide,
	TextField,
} from '@mui/material';
import { TransitionProps } from '@mui/material/transitions';

interface IProps {
	isAddingNewChat: boolean;
	onAddingModalClose: () => void;
	onAddComplete: (newChat: { name: string; chatAvatar: string }) => void;
}

const Transition = React.forwardRef(function Transition(
	props: TransitionProps & {
		children: React.ReactElement<any, any>;
	},
	ref: React.Ref<unknown>
) {
	return <Slide direction="up" ref={ref} {...props} />;
});

const AddNewChat = (props: IProps) => {
	const { isAddingNewChat, onAddingModalClose, onAddComplete } = props;
	const [chatName, setChatName] = useState('');
	const [avatarUrl, setAvatarUrl] = useState('');

	const onClickCreateBtn = () => {
		const newChat = { name: chatName, chatAvatar: avatarUrl };
		setChatName('');
		setAvatarUrl('');
		onAddComplete(newChat);
		onAddingModalClose();
	};
	return (
		<Dialog
			data-testid="wrapper"
			open={isAddingNewChat}
			TransitionComponent={Transition}
			keepMounted
			onClose={onAddingModalClose}
			aria-describedby="Adding new chat"
		>
			<DialogTitle>{'Adding new chat'}</DialogTitle>
			<DialogContent sx={{ width: '600px' }} dividers>
				<Box
					display="flex"
					flexDirection="column"
					alignContent="stretch"
					width="100%"
				>
					<TextField
						inputProps={{ 'data-testid': 'chat-name-input' }}
						value={chatName}
						onChange={(e) => {
							setChatName(e.target.value);
						}}
						id="Chat name"
						label="Chat name"
						type="text"
						variant="standard"
					/>
					<Box display="flex" height="100px" alignItems="center">
						<Avatar src={avatarUrl} data-testid="avatar" />
						<TextField
							inputProps={{ 'data-testid': 'avatar-url-input' }}
							value={avatarUrl}
							onChange={(e) => {
								setAvatarUrl(e.target.value);
							}}
							sx={{ flexGrow: 1, ml: 1 }}
							id="avatar-input"
							label="Chat avatar"
							type="text"
							variant="standard"
						/>
					</Box>
				</Box>
			</DialogContent>
			<DialogActions>
				<Button
					data-testid="close-btn"
					color="error"
					variant="outlined"
					onClick={onAddingModalClose}
				>
					Close
				</Button>
				<Button
					data-testid="create-btn"
					disabled={chatName.trim().length <= 2}
					color="success"
					variant="outlined"
					onClick={onClickCreateBtn}
				>
					Create
				</Button>
			</DialogActions>
		</Dialog>
	);
};

export default AddNewChat;
