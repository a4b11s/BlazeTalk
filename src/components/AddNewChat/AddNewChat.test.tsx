import React from 'react';

import {render, screen, waitFor} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import AddNewChat from './AddNewChat';

describe('AddNewChat component', () => {
    test('should be show when isAddingNewChat = true', () => {
        render(
            <AddNewChat
                isAddingNewChat={true}
                onAddingModalClose={() => {
                }}
                onAddComplete={() => {
                }}
            />
        );
        expect(screen.getByTestId('wrapper')).toHaveStyle('visibility: visible;');
    });
    test('should be hidden when isAddingNewChat = false', () => {
        render(
            <AddNewChat
                isAddingNewChat={false}
                onAddingModalClose={() => {
                }}
                onAddComplete={() => {
                }}
            />
        );
        expect(screen.getByTestId('wrapper')).toHaveStyle('visibility: hidden;');
    });
    test('avatar preview should be work', async () => {
        render(
            <AddNewChat
                isAddingNewChat={false}
                onAddingModalClose={() => {
                }}
                onAddComplete={() => {
                }}
            />
        );
        await waitFor(() => {
            // eslint-disable-next-line testing-library/no-wait-for-side-effects
            userEvent.type(screen.getByTestId('avatar-url-input'), 'test url');
        });
        expect(screen.getByRole('img', {hidden: true})).toHaveAttribute(
            'src',
            'test url'
        );
    });
    test('create button should be disable if the name contains less than 3 letters', async () => {
        render(
            <AddNewChat
                isAddingNewChat={true}
                onAddingModalClose={() => {
                }}
                onAddComplete={() => {
                }}
            />
        );
        await waitFor(() => {
            // eslint-disable-next-line testing-library/no-wait-for-side-effects
            userEvent.type(screen.getByTestId('chat-name-input'), 'te');
        });
        expect(screen.getByTestId('create-btn')).toBeDisabled();
    });
    test('create button should be disable if the name contains only white space', async () => {
        render(
            <AddNewChat
                isAddingNewChat={true}
                onAddingModalClose={() => {
                }}
                onAddComplete={() => {
                }}
            />
        );
        await waitFor(() => {
            // eslint-disable-next-line testing-library/no-wait-for-side-effects
            userEvent.type(screen.getByTestId('chat-name-input'), '   ');
        });
        expect(screen.getByTestId('create-btn')).toBeDisabled();
    });
    test('create button should not be disable if the name contains more than 3 letters', async () => {
        render(
            <AddNewChat
                isAddingNewChat={true}
                onAddingModalClose={() => {
                }}
                onAddComplete={() => {
                }}
            />
        );
        await waitFor(() => {
            // eslint-disable-next-line testing-library/no-wait-for-side-effects
            userEvent.type(screen.getByTestId('chat-name-input'), 'tes');
        });
        expect(screen.getByTestId('create-btn')).not.toBeDisabled();
    });
    test('onAddingModalClose should be call', async () => {
        const onAddingModalClose = jest.fn();
        render(
            <AddNewChat
                isAddingNewChat={true}
                onAddingModalClose={onAddingModalClose}
                onAddComplete={() => {
                }}
            />
        );
        await waitFor(() => {
            // eslint-disable-next-line testing-library/no-wait-for-side-effects
            userEvent.click(screen.getByTestId('close-btn'));
        });
        expect(onAddingModalClose).toBeCalled();
    });
    test('onAddComplete should be call with correct data', async () => {
        const testData = {name: 'test chat name', chatAvatar: 'test avatar url'};
        const onAddComplete = jest.fn();
        render(
            <AddNewChat
                isAddingNewChat={true}
                onAddingModalClose={() => {
                }}
                onAddComplete={onAddComplete}
            />
        );
        await waitFor(() => {
            // eslint-disable-next-line testing-library/no-wait-for-side-effects
            userEvent.type(screen.getByTestId('chat-name-input'), testData.name);
        });
        await waitFor(() => {
            // eslint-disable-next-line testing-library/no-wait-for-side-effects
            userEvent.type(screen.getByTestId('avatar-url-input'), testData.chatAvatar);
        });
        await waitFor(() => {
            // eslint-disable-next-line testing-library/no-wait-for-side-effects
            userEvent.click(screen.getByTestId('create-btn'));
        });
        expect(onAddComplete).toBeCalledWith(testData);
    });
    test('should to match snapshot', () => {
        const view = render(
            <AddNewChat
                isAddingNewChat={true}
                onAddingModalClose={() => {
                }}
                onAddComplete={() => {
                }}
            />
        );
        expect(view).toMatchSnapshot();
    });
});
