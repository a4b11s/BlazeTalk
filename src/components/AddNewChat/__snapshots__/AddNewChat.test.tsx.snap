// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNewChat component should to match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
      data-testid="wrapper"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
        role="presentation"
        style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
        tabindex="-1"
      >
        <div
          aria-describedby="Adding new chat"
          aria-labelledby=":r0:"
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-1t1j96h-MuiPaper-root-MuiDialog-paper"
          role="dialog"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
            id=":r0:"
          >
            Adding new chat
          </h2>
          <div
            class="MuiDialogContent-root MuiDialogContent-dividers css-1c30ccy-MuiDialogContent-root"
          >
            <div
              class="MuiBox-root css-1asqtti"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="Chat name"
                  id="Chat name-label"
                >
                  Chat name
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-v4u5dn-MuiInputBase-root-MuiInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                    data-testid="chat-name-input"
                    id="Chat name"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="MuiBox-root css-1a1tge1"
              >
                <div
                  class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
                  data-testid="avatar"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                    data-testid="PersonIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiFormControl-root MuiTextField-root css-16k1w2y-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for="avatar-input"
                    id="avatar-input-label"
                  >
                    Chat avatar
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-v4u5dn-MuiInputBase-root-MuiInput-root"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                      data-testid="avatar-url-input"
                      id="avatar-input"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-180lf2n-MuiButtonBase-root-MuiButton-root"
              data-testid="close-btn"
              tabindex="0"
              type="button"
            >
              Close
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeMedium MuiButton-outlinedSizeMedium Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-sxix9q-MuiButtonBase-root-MuiButton-root"
              data-testid="create-btn"
              disabled=""
              tabindex="-1"
              type="button"
            >
              Create
            </button>
          </div>
        </div>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
